### DISCUSSION TITLE
[Guide] How to make `goto()` and other movement instructions more reliable?

### DISCUSSION THREAD

**Discussion by MrSimbax** [colobot/colobot#1402] [Workshop]
Tired of getting random `Inaccesible location` errors? Want your CBOT program to work at 256x simulation speed *every* time? [Don't we all?](https://github.com/colobot/colobot/issues/336)

There's still hope. There are workarounds which might make your `goto()` experience better even on higher (greater than 1) simulation speeds. The workarounds could also be applied to other movement instruction like `move()` and `turn()`.

Most of this post is a list of ideas for workarounds. You may want to use some of them, mix them together and so on. I don't think there's one _general_ workaround, the solution will depend on your particular needs, the level, what your program is doing, what bot your program is for, the weather outside and so on.

## TL;DR (Example)

The following function appears to work well for a Code Battle AI. Copy and paste this function into your program then just use this function instead of `goto()`. Remember that it is just an example though and may not work well for your program or it might simply be an overkill.

```cpp
void trygoto(point pos)
{
	errmode(0);
	int tries = 0;
	while (goto(pos) != 0)
	{
		if(tries > 7)
		{
			goto(space());
			goto(pos, 0, 1);
			break;
		}
		else wait(2);
		tries++;
	}
	errmode(1);
}
```

## First step: turn off errors

Before we can do anything, we need to somehow stop a `goto()` error from terminating our program.

```cpp
errmode(0); // turn off errors
goto(pos); // will fail silently
errmode(1); // turn on errors back
```

This doesn't help us much by itself of course, the error is still there we're just ignoring it and pretending like nothing bad happened. The program will most definitely fail later since, surprise, surprise, the bot is not where we're expecting it to be. However, we now have the opportunity to detect the error and react accordingly.

It is recommended to use `errmode(1)` after you're done with `goto` so other parts of the program will still fail "loudly" so that you can immediately detect some bugs in the rest of your code.

From now on, the code snippets will assume `errmode(0)` was executed.

## Try again

Did you know that `goto()` returns a number? Well, now you [know](http://colobot.info/SatComOnline/master/cbot/goto).
The idea is to try `goto()` multiple times. It isn't that stupid if you believe that Colobot is non-deterministic (personally, I think it is).

```cpp
if (goto(pos) != 0)
{
	goto(pos);
}
```

If you want to try again more than once then use a loop.

```cpp
int tries = 0;
while (goto(item.position) != 0)
{
	if (tries >= 5) // let's try no more than 5 times
	{
		break;
	}
	tries++;
}
```

## Waiting between tries

It's probably a good idea to wait a little before trying again. This way we give the game more time to change state. Who knows, maybe the AlienAnt moving 2 centimeters forward on the other end of the map and the alignment of stars on the sky was the reason for the failing `goto()`?

```cpp
if (goto(pos) != 0)
{
	wait(2);
	goto(pos);
}
```

## Change your position

We can't easily change the position of the Titanium at the destination or the area around it without cheating but we can change our position and this is probably what matters the most for the path finding algorithm. Maybe something "invisible" is blocking the way from your current position? Maybe inaccurate calculations are at fault? Who knows. (It could also be the amount of radiation in your room or the temperature outside).

```cpp
goto(space());
goto(pos);
```

## Change your destination

Move close to the target and then align yourself properly "manually". For example, let's say that you want to grab a Titanium at `pos`:

```cpp
goto(space(pos));
turn(direction(pos));
move(distance(this.position, pos) - 3.0);
grab();
```

## Pass more parameters

`goto()` actually accepts more than just the position which allows to modify the behavior of the path finding algorithm a little. Example:

```cpp
goto(pos, 0, 1);
```

This tells `goto()` that the bot should stay on the ground and that getting _close_ to the destination is enough, we don't need to reach precisely the goal. Refer to [`SatCom`](http://colobot.info/SatComOnline/master/cbot/goto) for more information.

## More

Share your ideas for workarounds for the nasty `goto()` function or your `trygoto()` functions in the comments. Keep in mind I might "steal" some of the ideas and put them in the guide so that all the known methods are aggregated in one place for easy reference, so I hope you don't mind.

**Comment by SimonHeimberg** [comment]
What is also helpful are the two variants of `try ... catch`.

#### error catching
One variant of `try ... catch` is for catching errors (like `catch` in other programming languages).

```cpp
try {
    goto(somePosition);
    grab();
    ...
} catch (102 /*ERR_MANIP_BUSY == already holds something*/) {
    turn(20);
    drop();
} catch (212 /*ERR_GOTO_BUSY == goto destination occupied*/) {
    // could try to find out what is in the way, and remove it when it is something portable
    ...
} catch(CBotErrDeletedPtr /*Attempted to use deleted object */) {
    ...
} catch(CBotErrNull /* Attempted to use a null pointer */) {
   ...
} finally {
   // is always run
}
```
Find the numeric error codes in https://github.com/colobot/colobot/blob/d4ab82eaae32fa3a68d0699a941db67bc7fb9c18/src/common/error.h#L31 or by getting the error code like this:
```cpp
errmode(0);
int err = goto(somePositon); // expected to fail
errmode(1);
if (err != 0) {
    message("failed with err " + err, DisplayInfo);
}
```
Find the error constants (CBotErrDeletedPtr, ...) in https://github.com/colobot/colobot/blob/d4ab82eaae32fa3a68d0699a941db67bc7fb9c18/src/CBot/CBotProgram.cpp#L394 (you can also use the numeric value instead of the constant).

You can throw your own error code:
```cpp
try {
    ...
    throw(81001); // must be > 0
    ...
} catch(81001) {
    // handle own error
}
```

Maybe I first learnt about catching errors in a comment in https://github.com/colobot/colobot/issues/474

For me `try ... catch` is very useful to use in the main loop. Then the program can retry to do something instead of aborting. But when the error happens in every iteration, it is difficult to find out where its origin is.
```cpp
while(true) {
    try {
        ...
        lastErr = 0;
    } catch(211 /* ERR_GOTO_ITER  ==  goto too complicated */) {
        if (211 == lastErr) {
            // do something witch could improve the situation, like goto() for a part of the way
        }
        message("Error 211, goto too complicated", DisplayWarning);
        lastErr = 211;
        ...
    }
}
```

Advantages over `errmode(0)`:
* you can not forget to do `errmode(1)`
* you can use it for a small or big portion of the code.

Disadvantages over `errmode(0)`:
* you must catch each error individually

Things to not forget:
* When using it on a big code portion, there is no simple possibility to find out where exactly the error was thrown.
  (At least I do not know any error function or variable.) One way is to temporarily change the code to catch a number which is never thrown (like 999102 instead of 102). But maybe the problem will not occur anymore after stop and start...
* After `errmode(0)`, no errors are thrown by functions (like `goto()` or `grab()`) as normally.

**Comment by SimonHeimberg** [comment]
#### event triggering
The other variant of `try ... catch` is something like event triggering. For this, you put a **condition** in `catch(...)`, which is checked regularly.
```cpp
try {
    goto(somePosition);
} catch(null != this.energyCell && this.energyCell.energyLevel < 0.1 ) {
    // sorry, does NOT work, see comment below
    rechargeBattery();
}
```

I found this on https://colobot.info/did_you_know/theres-a-mysterious-try-catch-block-statement-in-cbot-that-allows-you-to-interrupt-normally-non-interruptible-instructions-like-move-and-goto/

- [ ] Sorry, aborting `goto()` does NOT work, as long as #474 is not fixed. The goto() command is completed first, even if other commands are given. Triggering an event could still be useful for other commands, but I did not find a use case yet.

Limitting the time for some operation should work *in the feature* like this:
```cpp
int timeAbort = abstime() + timeWait;
try {
	goto(endTarget, height, 1 /* ~noStop */);
} catch ( abstime() > timeAbort) {
	// time limit reached
	// sorry, does NOT work, see comment above
	...
}
```

**Comment by SimonHeimberg** [comment]
#### Pass more parameters to goto()
goto() even has four parameters, see in SatCom

The values for the parameters are defined in 
https://github.com/colobot/colobot/blob/d4ab82eaae32fa3a68d0699a941db67bc7fb9c18/src/object/task/taskgoto.h#L39

The 3rd parameter `goal` is what to do when at the target (or close by??) and is documented in SatCom. (0 = stop, 1 = express (without stopping), -1 = default = same as stop) Maybe this also has an effect on a collision. (At least sometimes my bot picks up wrong things.) My understanding of one use case: When you do several goto commands directly after each other (following some path), set this to 1.

And the 4th parameter `crash` is what do on crash (0 = stop, 1 = right-left, 2 = left-right, 3 = left, 4 = right, 5 = beam ??, -1 = default = same as beam) Not sure how to use it and what the values mean. But it is used (as value 1) in the program which is submitted by Houston for the shielder in "TerraNova" Level 4 (the queen).
Maybe somebody is successful in understanding https://github.com/colobot/colobot/blob/d4ab82eaae32fa3a68d0699a941db67bc7fb9c18/src/object/task/taskgoto.cpp#L649 (I found there what the default values are used as.)

**Comment by SimonHeimberg** [comment]
#### general: check the error of goto()
To improve failures of `goto()`, the error should be checked. No matter if with `errmode(0)` or with `try ... catch() ...`.
Moveing itself around may help in many cases, but not really when the goto destination is occupied.

```cpp
errmode(0);
int err = goto(dest);
errmode(1);
if (0 == err) {
    return; // passed
} else if (210 == err /* goto impossible */) {
    // How can this be improved? It sounds very general.
} else if (211 == err /*goto too complicated*/) {
    // got in the direction of the target, or choose an other target of the same type, or ...
} else if(212 == err /* goto destination occupied */) {
    // maybe goto near by and check if it is free now (the other bot might have driven away)
    //     if still not free, then check what could be the blocking object and remove it if possible
} else {
    // Should not happen normally. Maybe CBotErrDeletedPtr when you do `goto(someObject.position)` and the object disappeared just before (built something from a titanium, converted, destroyed)
}
```

```cpp
try {
    goto(dest);
} catch(211 /*goto too complicated*/) {
        // got in the direction of the target
} catch (...) {
    ... // as above for errmode(0)
```

**Comment by SimonHeimberg** [comment]
#### general hints

* space(position, rmin, rmax, dist) has a optional parameters. Use them if it looks helpful.
* move() allows to drive into water. So only use it for short movements and stick to goto() for longer movements. Or check yourself if there is water (`topo(posBetweenYouAndTarget) < -0.4`, maybe 0.4 m of water is already too deep, not sure what knee deep is in meters.).
* when you stop trying, indicate this somehow to stop the normal program flow. (`trow 10002;`, `throw err;`, `return false;`, `return err;`, `errmode(1); goto(dest); ` to show the usual error message, ...

**Comment by gitgublu** [comment]
About water, the `goto()` function don't verify the altitude before land the bot when its jet overheats. So, sometimes, when the bot is over water, `goto()` land it into water and the bot explodes, that is a little disappointing... To resolve this problem, the `goto()` function should be modified... Or, we can create a similar function ! So, I tried, but the goto function is very complex and probe the lengh of a body of water is not simple... I have this basic function that assume that the body of water is traversable, but that cool down the jet before:
```
public void object::newgoto(point p){
	bool b=true;
	
	while((this.position.z-topo(this.position))<5){
		jet(0.3);
	}
	while(distance2d(this.position,p)>1){
		turn(direction(p));
		motor(0.5,0.5);
		try{
			if(topo(this.position)<0){
				while(this.position.z<5){
					jet(0.3);
				}
				while(this.position.z>5){
					jet(-0.3);
				}
			}
			else{
				b=true;
				while((this.position.z-topo(this.position))<5){
					jet(0.3);
				}
				while((this.position.z-topo(this.position))>5){
					jet(-0.3);
				}
				if(this.temperature>0.7){
					jet(-0.3);
					while((this.position.z-topo(this.position))>0);
					jet(0);
					while(this.temperature>0);
				}
			}
		}catch((topo(this.position)<=0)&&(this.temperature>0.2)&&b){
			jet(0);
			motor(-0.5,-0.5);
			while(topo(this.position)<=0);
			motor(0,0);
			jet(-0.3);
			while((this.position.z-topo(this.position))>0);
			jet(0);
			while(this.temperature>0);
			jet(0.3);
			while((this.position.z-topo(this.position))<5);
			motor(1,1);
			while(topo(this.position)>=0);
			b=false;
		}
	}
	motor(0,0);
	jet(-0.3);
	while((this.position.z-topo(this.position))>0);
	jet(0);
 }
```
it a draft, so I'll improve this code and I'll up it to date.

**Reply by gitgublu** [reply]
When "[Try...Catch doesn't interrupt goto()](https://github.com/colobot/colobot/issues/474)"'ll be fixed, it will be possible to reuse the goto function in a `try...catch` bloc where the catch condition will be the presence of water forward. So, it needs to use vectors, and the Thales theorem:
![image](https://user-images.githubusercontent.com/95926527/199781889-716148ba-379d-4d16-991c-1059eb8f6105.png)
code will can be like:
```

// test function
public bool object::probeground(){
   float d=sqrt(pow(this.position.x,2)+pow(this.position.y,2));
   point a (this.position.x*((d+5)/d),this.position.y*((d+5)/d),0) //altitude is not important to probe ground height
   return (topo(a)<=0);
}

//your program in extern void object::Yourprogram(){...}
while(distance(this.position, somewhere)>1){
    try{
       goto(somewhere);
    }
    catch(probeground()){
        //verify that the body of water is traversable, else cool down the jet or throw an error if it is still too lengh
    }
}
```

**Comment by kompowiec** [comment]
rgoto https://github.com/colobot/colobot/discussions/1376
